### PR TITLE
Make group delete async

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**: Follow up on #1610, this PR enables async reconcile for groups. See #1541  for the proposal.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

/hold

will rebase on top of #1610 once it merges.

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Make group delete async
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1667/commits/de4e1992a80aa4a10f786b8db3d1b84ceeeea4da is the relevant commit for this PR (the other changes are already part of #1610)

File: api/v1alpha4/conditions_consts.go
Comment (CecileRobertMichon) [inline api/v1alpha4/conditions_consts.go:24]: re-added so that the cluster doesn't show as Ready until all services are done reconciling (might remove once each service has its own condition in place to use as summary)

File: azure/services/groups/groups.go
Comment (shysank) [inline azure/services/groups/groups.go:78]: We should think about handling the cancellation signal here. Right now, we are just cleaning up the resources, but we should also think about what will happen if client.Get for eg. takes longer than "DefaultAzureServiceReconcileTimeout".  It's not going to break anything, but we are not returning even if it exceeds the timeout which was the original intention. wdyt?

File: controllers/azurecluster_controller.go
Comment (CecileRobertMichon) [reply controllers/azurecluster_controller.go:162]: the Ready condition gets set as the summary as part of PatchObject() which is called at the end of every loop https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1667/files#diff-b74c6acb4629abe58f8286da9939563cc4404e09fe9966eb61dbde9b87274189R553

this commit reverts part of #1610 which removed NetworkInfrastructureReadyCondition so that we have a generic condition for everything that isn't async yet so that the summary doesn't set Ready as "true" until all the resources are actually ready (Summary also takes NetworkInfrastructureReadyCondition as one of the conditions so Ready is true when both NetworkInfrastructureReadyCondition and all the individual resource conditions are True)

File: azure/services/groups/groups.go
Comment (CecileRobertMichon) [reply azure/services/groups/groups.go:78]: > but we are not returning even if it exceeds the timeout which was the original intention

can you elaborate on this? 

maybe https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1684/files#diff-5bb14c98c52ecfa3dad6cf4745c9378b7624fdc60f0910343534ddac18408fb1R62 is a better illustration of how this works not just for creates... if the context expires in this function it will return an error which will cause that reconcile loop to fail. For long-running operations, like deletes and creates, we expect that to happen which is why we return a transient "operation not done" error which causes a requeue. If the overall loop times out, eg. a Get takes longer than it should, then maybe it's a good thing to error out and retry later? This is meant to be as an upper bound, ie. this service should never take more than 12 seconds to run the Delete/Reconcile function, unless something went wrong.

File: controllers/azurecluster_controller.go
Comment (shysank) [reply controllers/azurecluster_controller.go:162]: Got it, Thanks for the explanation.

File: azure/services/groups/groups.go
Comment (shysank) [reply azure/services/groups/groups.go:78]: > if the context expires in this function it will return an error which will cause that reconcile loop to fail

Not sure if that'll happen, though. We'll need to watch for `ctx.Done()` to get the cancellation signal, and return error accordingly. Something similar to what `future.WaitForCompletionRef` does [here](https://github.com/Azure/go-autorest/blob/492f94bc3a562197e229f044dbfabcc0ecb2489a/autorest/azure/async.go#L252). Does that make sense?

Comment (CecileRobertMichon) [reply azure/services/groups/groups.go:78]: I see what you're saying now. I guess this isn't really a long-running function where we want to continuously do something and timeout though. The reason for doing a cxt timeout here is that the collection of all the shorter timeouts should never exceed this bigger timeout. When go sets a new timeout deadline, it will select the soonest between the parent context's deadline and the new deadline, so this is meant more as an intermediate timeout (in this case it doesn't really apply though because the group delete only performs one delete operation so it's a 1:1 mapping), same way that the reconcile loop timeout is a global timeout of all the long-running operations we have. Maybe we can discuss this at office hours tomorrow it will be easier to talk it through.

Comment (shysank) [reply azure/services/groups/groups.go:78]: Ah I see, so this timeout is just to make sure that the `time(all async operations) < DefaultReconcileTimeout` , and not necessarily `time(service.Reoncile/Delete) < DefaultReconcileTimeout`. I think it makes sense, there shouldn't be any network calls outside of the async interface. ResourceGroups is a special case I guess to check for `IsManaged`, and for other services this will be picked up from a cache.

**Comment by CecileRobertMichon** [comment]
would be great to get a second lgtm on this if anyone has time to put some 👀  on it 

@nader-ziada @devigned @whites11

**Comment by CecileRobertMichon** [comment]
I changed the service name from "groups" to "group" so it reads "group deleting" instead of "groups deleting" in the logs (there is always only one group per cluster)

File: azure/services/groups/client.go
Comment (devigned) [inline azure/services/groups/client.go:131]:
```suggestion
        ctx, span := tele.Tracer().Start(ctx, "groups.azureClient.resourceGroupParams")
	defer span.End()
	
	var params interface{}
```

File: azure/services/groups/spec.go
Comment (devigned) [inline azure/services/groups/spec.go:54]: What if additional tags were added?

**Review by devigned — CHANGES_REQUESTED** [review]
Just a couple suggestions. Overall looks great!

File: azure/services/groups/spec.go
Comment (nader-ziada) [reply azure/services/groups/spec.go:54]: I don't think it would update currently

**Comment by nader-ziada** [comment]
lgtm, pending David's comments

File: azure/services/groups/spec.go
Comment (CecileRobertMichon) [reply azure/services/groups/spec.go:54]: indeed, but we weren't doing that before either: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1667/files#diff-036de5e90528343ff92467735b06ba4c22abbb0616cc3911820360bc37aea755L59 

I tried to avoid making any sort of bug fixes / behavioral changes (besides async stuff) as part of these PRs because they are already so big. How about I file a bug issue to fix it separately?

File: azure/services/groups/client.go
Comment (CecileRobertMichon) [reply azure/services/groups/client.go:131]: do we want to be starting spans for functions that don't make any outside calls ? this function is just a getter so execution should be quick.

File: azure/services/groups/spec.go
Comment (devigned) [reply azure/services/groups/spec.go:54]: sgtm

File: azure/services/groups/client.go
Comment (devigned) [reply azure/services/groups/client.go:131]: I'd agree, but just below this is `existingRG, err := ac.Get(ctx, spec.ResourceName())`, which does make a call. If you were looking at the spans, you would see the parent span come from the grandparent func. The spans would not tell the code story as well if you don't fill in the span names all the way down. WDYT?

Comment (CecileRobertMichon) [reply azure/services/groups/client.go:131]: oh my bad, I thought this comment was on the paramaters func in spec.go

you're totally right, this deserves a span, adding it

Comment (CecileRobertMichon) [reply azure/services/groups/client.go:134]: added

File: azure/services/groups/spec.go
Comment (CecileRobertMichon) [reply azure/services/groups/spec.go:54]: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1696

**Review by devigned — APPROVED** [review]
/lgtm
/approve

**Comment by CecileRobertMichon** [comment]
/retest

windows vmss flake

**Pull request merged**
